src/libsystemd/sd-netlink/generic-netlink-message.c

```diff
@@ -0,0 +1,267 @@
+        _cleanup_close_ int fd = -1;
+        int r;
+
+        fd = socket_open(NETLINK_GENERIC);
+        if (fd < 0)
+                return fd;
+
+        r = sd_netlink_open_fd(ret, fd);
+        if (r < 0)
+                return r;
+
+        fd = -1;
+
+        return 0;
+}
+
+int sd_rtnl_message_new_generic_netlink(uint16_t type, int command, int version, sd_netlink_message **ret) {
+        _cleanup_(sd_netlink_message_unrefp) sd_netlink_message *m = NULL;
+        struct genlmsghdr *g;
+        size_t size;
+        int r;
+
+        r = message_new_empty(NULL, &m);
+        if (r < 0)
+                return r;
+
+        size = NLMSG_SPACE(sizeof(struct genlmsghdr));
+        m->hdr = malloc0(size);
+        if (!m->hdr)
+                return -ENOMEM;
+
+        m->hdr->nlmsg_type = type;
+        m->hdr->nlmsg_flags = NLM_F_REQUEST | NLM_F_ACK;
+        m->hdr->nlmsg_len = size;
+
+        g = NLMSG_DATA(m->hdr);
+        g->cmd = command;
+        g->version = version;
+
+        *ret = m;
+        m = NULL;
+
+        return 0;
+}
+
+int sd_generic_netlink_message_append_string(sd_netlink_message *m, unsigned short type, const char *data) {
+        size_t length;
+        int r;
+
+        assert_return(m, -EINVAL);
+        assert_return(!m->sealed, -EPERM);
+        assert_return(data, -EINVAL);
+
+        length = strlen(data);
+
+        r = sd_netlink_message_append_data(m, type, data, length + 1);
+        if (r < 0)
+                return r;
+
+        return 0;
+}
+
+int sd_generic_netlink_message_append_u8(sd_netlink_message *m, unsigned short type, uint8_t data) {
+        int r;
+
+        assert_return(m, -EINVAL);
+        assert_return(!m->sealed, -EPERM);
+
+        r = sd_netlink_message_append_data(m, type, &data, sizeof(uint8_t));
+        if (r < 0)
+                return r;
+
+        return 0;
+}
+
+int sd_generic_netlink_message_append_u16(sd_netlink_message *m, unsigned short type, uint16_t data) {
+        int r;
+
+        assert_return(m, -EINVAL);
+        assert_return(!m->sealed, -EPERM);
+
+        r = sd_netlink_message_append_data(m, type, &data, sizeof(uint16_t));
+        if (r < 0)
+                return r;
+
+        return 0;
+}
+
+int sd_generic_netlink_message_append_u32(sd_netlink_message *m, unsigned short type, uint32_t data) {
+        int r;
+
+        assert_return(m, -EINVAL);
+        assert_return(!m->sealed, -EPERM);
+
+        r = sd_netlink_message_append_data(m, type, &data, sizeof(uint32_t));
+        if (r < 0)
+                return r;
+
+        return 0;
+}
+
+int sd_generic_netlink_message_parse_container(struct rtattr *rta, struct rtattr *tb[]) {
+        struct rtattr *rt;
+        uint32_t len;
+
+        assert_return(rta, -EINVAL);
+        assert_return(tb, -EINVAL);
+
+        rt = RTA_DATA(rta);
+        len = RTA_PAYLOAD(rta);
+
+        for (; RTA_OK(rt, len); rt = RTA_NEXT(rt, len))
+                tb[rt->rta_type] = rt;
```

Comment:
`CTRL_ATTR_MAX` is only for the control. We need to parse the attributes which is a variable size.
The caller should be knowing the size . For example ` ATTR types defined for L2TP' has `__L2TP_ATTR_MAX`  . We need to allocate the array size of that length.